tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/legalize_to_linalg.cc

```diff
@@ -224,6 +224,286 @@ static bool HasCanonicalDimensionNumbers(
+ public:
+  using OpConversionPattern<mhlo::EinsumOp>::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      mhlo::EinsumOp op, ArrayRef<Value> args,
+      ConversionPatternRewriter& rewriter) const final {
+    auto get_rank = [](Value v) {
+      return v.getType().cast<ShapedType>().getRank();
+    };
+    auto is_scalar = [&](Value v) { return get_rank(v) == 0; };
+    auto einsum_config = op.einsum_config();
+
+    // Setting Up Commonly Used Symbols as strings
+    const std::string kArrow = "->";
+    const std::string comma = ",";
+    const std::string ellipsis = "...";
+
+    // With the assumption of binary Input Operand and Single Output
+    // Get the inputs and output operands' indices
+    // einsum_config = "lhs_loop,rhs_loop->output_loop"
+    std::size_t pos_kArrow = einsum_config.find(kArrow);
+    std::size_t pos_comma = einsum_config.find(comma);
+
+    StringRef lhs_loop = einsum_config.substr(0, pos_comma);
+    StringRef rhs_loop = einsum_config.substr(
+        pos_comma + comma.size(), pos_kArrow - (pos_comma + comma.size()));
+    StringRef output_loop = einsum_config.substr(pos_kArrow + kArrow.size());
+
+    // Check for Invalid Configs
+    // 1.Check that there is only maximum 2 inputs
+    // 2.Check that there is only maximum 1 output
+    // 3.Check that there is 1 kArrow
+    if (rhs_loop.find(comma) != std::string::npos ||
+        output_loop.find(comma) != std::string::npos ||
+        output_loop.find(kArrow) != std::string::npos) {
+      return rewriter.notifyMatchFailure(op, "Invalid Einsum Config!");
+    }
+
+    // Find result type, if on tensors.
+    ShapedType result_ty;
+    result_ty = this->typeConverter->convertType(op->getResultTypes().front())
+                    .template dyn_cast<ShapedType>();
+
+    // Check result type compatibility.
+    if (!result_ty || !result_ty.hasRank() ||
+        !(result_ty.getElementType().isSignlessIntOrFloat() ||
+          result_ty.getElementType().isa<ComplexType>())) {
+      return rewriter.notifyMatchFailure(
+          op, "mismatched operand/result types or iterator count");
+    }
+
+    // Convert the representation from string/vector<char> to vector<string>
+    // i.e ("abc") -> {"a", "b", "c"} S.T we can have no limit on number of
+    // loops Since ASCII Chars has only 256 number of characters Find elilipsis
+    // in loop and replace and find number of dim/rank ellipsis represent then
+    // add batch dims into loop_dim as loop_dim.push_back({"0",... "N"}) where N
+    // is rank of batch then we can add the remaining dimensions. For example
+    // given config "ab...cde" representing where batch rank is 3 then we will
+    // get loop_dim = f("ab...cde") = {"a","b","0","1","2","c","d","e"}
+    SmallVector<SmallVector<std::string>> all_loop_dim;
+    SmallVector<int, 3> batch_rank_vec;
+    size_t max_operand_sz = 3;  // 2 Inputs + 1 output operand
+    for (auto loop : llvm::enumerate(
```

Comment:
It looks like the whole for-loop is trying to build batch_rank?

If so, let's make it a function and call it three times for lhs, rhs, out.

Reply:
It's to get batch_rank and represent the einsum config as vector<string>.
Should I split it up to two different blocks and call out three times in each block to make it look cleaner

i.e
lhs_vec_ein = GetEinVecRepr(lhs);
rhs_vec_ein = GetEinVecRepr(lhs);
out_vec_ein = GetEinVecRepr(lhs);

lhs_batch_rank = GetBatchRank(lhs):
rhs_batch_rank = GetBatchRank(rhs):
out_batch_rank = GetBatchRank(out):

Or perhaps should I make a function that return a pair containing (Einsum_config, batch_rank)/ pair<Vector<string>,Int>

Reply:
I am wondering if it's better to switch back to representing it as vector<char>/string instead of vector<string>, as seen [here](https://github.com/raikonenfnu/iree/commit/e81d9106cfeeb219bebfee0c21b56e4ed47a8733#diff-22257ffd15247f478e895da2e07bcd9a9dd498b43765b05f0aec40affdd4a080R434), as the loop/function becomes much simpler and straight forward. With caveat that since there exist only limited amount of char, number of dimension representable is also limited(there exist 256 ascii chars => 256 dimensions/ranks). But I was thinking maybe that is plenty enough for most people's use case.

Reply:
not sure how others think. Splitting it up to two different blocks and calling out three time looks good to me.

Reply:
done